### PR TITLE
Mark some Halider::Buffer() ctors explicit

### PR DESCRIPTION
**Pull request by steven-johnson**
Buffer(buffer_t) and Buffer(halide_buffer_t) should be marked 'explicit' to avoid quiet auto-conversion; since these can fail with a runtime error if the argument's size/type is a mismatch with the Buffer<> static type declaration, this can quietly inject runtime failures when converting code to use Buffer<>.

### CODE REVIEW

**Comment by steven-johnson** [comment]
Also added `Buffer(buffer_t, halide_type_t)` ctor, to allow specifying the type at runtime.
Also made the check in `Buffer(buffer_t)` that T is not void into a static assert, since it's never OK to create a buffer this way (we wouldn't be able to set a type correctly).

File: src/runtime/HalideBuffer.h [hunk 1]
```diff
@@ -562,28 +582,22 @@ class Buffer {
...
-        make_shape_storage();
-        for (int i = 0; i < d; i++) {
-            buf.dim[i].min = old_buf.min[i];
-            buf.dim[i].extent = old_buf.extent[i];
-            buf.dim[i].stride = old_buf.stride[i];
-        }
-        buf.set_host_dirty(old_buf.host_dirty);
-        assert(old_buf.dev == 0 && "Cannot construct a Halide::Runtime::Buffer from a legacy buffer_t with a device allocation. Use halide_upgrade_buffer_t to upgrade it to a halide_buffer_t first.");
+    /** Make a Buffer from a legacy buffer_t, with an explicit halide_type. */
+    explicit Buffer(const buffer_t &old_buf, halide_type_t t) {
```
Comment (abadams) [inline src/runtime/HalideBuffer.h]: Makes sense that you can't use the constructor below with ```Buffer<void>```. Thus it seems weird that you *can* use *this* constructor with anything *other* than ```Buffer<void>```

**Comment by abadams** [comment]
Sigh, looks like opencl is broken on the linux bots

File: src/runtime/HalideBuffer.h (hunk 1 above)
Comment (steven-johnson) [reply src/runtime/HalideBuffer.h]: You can use it when the dynamic type exactly matches the static type. Not a likely use case, but it is possible, so I left it.

**Pull request merged**